### PR TITLE
enh: Optimize groupby nunique implementation

### PR DESCRIPTION
**Pull request by ChengjieLi28**
## What do these changes do?
Optimize groupby nunique implement.

## Check code requirements

- [x] tests added / passed (if needed)
- [ ] Ensure all linting tests pass, see [here](https://docs.pymars.org/en/latest/development/contributing.html#check-code-styles) for how to run them

### CODE REVIEW

File: mars/dataframe/groupby/nunique.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+    @classmethod
+    def get_execute_map_result(
+        cls, op: DataFrameGroupByAgg, in_data: pd.DataFrame
+    ) -> Union[pd.DataFrame, pd.Series]:
+        selections = cls._get_selection_columns(op)
+        by_cols = op.raw_groupby_params["by"]
+        if by_cols is not None:
+            cols = (
+                [*selections, *by_cols] if selections is not None else in_data.columns
```
Comment (aresnow1) [inline mars/dataframe/groupby/nunique.py:88]:
```suggestion
                selections + by_cols if selections is not None else in_data.columns
```

File: mars/dataframe/groupby/nunique.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+            )
+            res = in_data[cols].drop_duplicates(subset=cols).set_index(by_cols)
+        else:  # group by level
+            selections = selections if selections is not None else in_data.columns
+            level_indexes = cls._get_level_indexes(op, in_data)
+            in_data = in_data.reset_index()
+            index_names = in_data.columns[level_indexes].tolist()
+            cols = [*index_names, *selections]
```
Comment (aresnow1) [inline mars/dataframe/groupby/nunique.py:96]:
```suggestion
            cols = index_names + selections
```

**Review by aresnow1 — COMMENTED** [review]
LGTM overall, left some comments.

**Review by qianduoduo0904 — COMMENTED** [review]
I left some comments about a few details.

**Review by qianduoduo0904 — APPROVED** [review]
LGTM

**Review by aresnow1 — APPROVED** [review]
LGTM

**Pull request merged**